docs/internals/services.rst

```diff
@@ -1,6 +1,112 @@
-.. include:: /_shared/stub-topic.txt
+Understanding EF Services
+=========================
 
-|stub-icon| Understanding EF Services
-=====================================
+.. contents:: `In this article`
 
-.. include:: /_shared/stub-notice.txt
+Entity Framework executes as a collection of services working together. A
+service is a reusable component. In C#, a services is typically an
+implementation of an interface. Services are is available to other services via
+`dependency injection (DI) <https://wikipedia.org/wiki/Dependency_injection>`_.
+This article covers some fundamentals principals for understanding how EF uses
+services and DI.
+
+Categories of Services
+----------------------
+
+Services fall into one or more categories.
+
+Context services
+  Services that related to interaction with a specific instance of
+  ``DbContext``. These services provide functionality for working with the user
+  model and context options.
+
+Provider services
+  Provider-specific implementations of services. For example, SQLite uses
+  "provider services" to customize the behavior of SQL generation, migrations,
+  and file I/O as needed.
+
+Design-time services
+  Services used when a developer is creating an application. For example, EF
+  commands uses design-time services to execute migrations and code generation
+  (aka scaffolding).
+
+User services
+  A user can define custom services to interact with EF. These are written in
+  application code, not provider code. For example, users can provide an
+  implementation of ``IModelCustomizer`` for controlling how a model is created.
+
+
+Services Initialization
+-----------------------
+
+In order to be used by EF, all services must be registered in a service
+collection before or during ``DbContext`` initialization. Registration is done
+by adding components to into the ``ServiceCollection`` that EF will eventually
+use to create the final ``ServiceProvider``. By the time initialization is
+complete, ``DbContext`` has an instance of ``ServiceProvider`` that includes the
+collection of provider services, context services, and user services necessary
+for operation.
+
+.. note::
+  ``ServiceProvider`` is not to be confused with a "provider's services".
+
+Initialization Steps
+^^^^^^^^^^^^^^^^^^^^
+
+EF will initialize services in the following order. Services registered later
+can override or remove previously registered services.
+
+1. Start with the ``ServiceProvider`` injected in the DbContext constructor (*user* and *design-time services*). If no services are provided, create a new instance.
```

Comment:
We should consider having sections for:
- EF using an external vs. internal service provider 
- `AddDbContext<TContext>()`: 
 - How it is useful/when to use it
 - How it works, e.g. every service it sets up and how it uses `DbContextActivator` so that an `IServiceProvider` and `DbContextOptions` parameter is not required in the constructor to be injected during initialization.